gdpr/anonymizers/fields.py

```diff
@@ -0,0 +1,246 @@
+        deanonymize_func=lambda a, x: x**2-a.get_numeric_encryption_key()
+        )
+    ```
+    """
+
+    anon_func: Callable
+    deanonymize_func: Optional[Callable] = None
+    max_anonymization_range: int
+
+    def __init__(self,
+                 anon_func: Union[Callable[[Any], Any], Callable[["FunctionFieldAnonymizer", Any], Any]],
+                 deanonymize_func: Callable[["FunctionFieldAnonymizer", Any], Any] = None,
+                 *args, **kwargs) -> None:
+        super().__init__(*args, **kwargs)
+        if callable(anon_func):
+            self.anon_func = anon_func  # type: ignore
+        else:
+            raise ImproperlyConfigured('Supplied func is not callable.')
+
+        if callable(deanonymize_func):
+            self.deanonymize_func = deanonymize_func
+        elif deanonymize_func is not None:
+            raise ImproperlyConfigured('Supplied deanonymize_func is not callable.')
+        else:
+            self.is_reversible = False
+
+    def get_numeric_encryption_key(self) -> int:
+        return numerize_key(self.get_encryption_key()) % self.max_anonymization_range
+
+    def get_encrypted_value(self, value):
+        if self.deanonymize_func is None:
+            return self.anon_func(value)
+        else:
+            return self.anon_func(self, value)
+
+    def get_is_reversible(self, obj=None):
+        return self.deanonymize_func is not None
+
+    def get_decrypted_value(self, value):
+        if not self.get_is_reversible():
+            raise self.IrreversibleAnonymizationException()
+        else:
+            return self.deanonymize_func(self, value)
+
+
+class DateFieldAnonymizer(NumericFieldAnonymizer):
+    """
+    Anonymization for DateField.
+
+    """
+
+    max_anonymization_range = 365 * 24 * 60 * 60
+
+    def get_encrypted_value(self, value):
+        return value - timedelta(seconds=self.get_numeric_encryption_key())
+
+    def get_decrypted_value(self, value):
+        return value + timedelta(seconds=self.get_numeric_encryption_key())
+
+
+class CharFieldAnonymizer(FieldAnonymizer):
+    """
+    Anonymization for CharField.
+
+    """
+
+    def get_encrypted_value(self, value):
+        return encrypt_message(self.get_encryption_key(), value)
+
+    def get_decrypted_value(self, value):
+        return decrypt_message(self.get_encryption_key(), value)
+
+
+class EmailFieldAnonymizer(FieldAnonymizer):
+
+    def get_encrypted_value(self, value):
+        return encrypt_email(self.get_encryption_key(), value)
+
+    def get_decrypted_value(self, value):
+        return decrypt_email(self.get_encryption_key(), value)
+
+
+class DecimalFieldAnonymizer(NumericFieldAnonymizer):
+    """
+    Anonymization for CharField.
+    """
+
+    max_anonymization_range = 10000
+    decimal_fields: int = 2
+
+    def get_encrypted_value(self, value):
+        return value + Decimal(self.get_numeric_encryption_key()) / Decimal(10 ** self.decimal_fields)
```

Comment:
proč to dělíš? Můžeš tuhle anonimizaci popsat? Co když value už desetinné je. Jde ti o to aby desetinné místo nezůstalo?

Reply:
Tady `self.get_numeric_encryption_key()` vždy vráti celé číslo. Abychom mohli nějak naonymizovat i desetinou část tek pomocí `Decimal(self.get_numeric_encryption_key()) / Decimal(10 ** self.decimal_fields)` z `Decimal(314159265)` a `decimal_fields=2` se udělá `Decimal(3141592.65)`